Introspection sources improvements, part 2 (#18277)
MaterializeInc/materialize: doc/user/content/ops/troubleshooting.md

Author: As part of #17563 we should probably do a more holistic treatment of this document. I consider that out-of-scope for this PR, but these are the things I notices:

* There is nothing about inspecting dataflow frontiers here, which tends to be hugely important for debugging bad query performance in cloud.
* Most sections include both per-worker and aggregated queries. At least some of the per-worker queries can probably be omitted.
* The sections talking about subscribes were written before `mz_subscriptions` existed. I assume these queries can now be significantly smaller.

issues with Materialize.

{{< warning >}}
The queries below reference [introspection relations](/sql/system-catalog/mz_internal/#replica-introspection-relations), which are subject to future changes without notice.

> This is why these relations are all created in the `mz_internal` schema,
> and why it is not possible to create higher-level views depending on them.

Reviewer: ```suggestion
All relations below are created in the `mz_internal` schema and it is not possible to create higher-level views dependent on them.
```

Author: The thing we'd like to convey here is "sorry, you can't create persistent views for these troubleshooting queries because the relations they read from are unstable". So it would be nice to include a reference to the previous paragraph somehow (like the "this is why...").

Reviewer: I understand the context now, thanks for the explanation. Please check out the new suggestion and make sure it conveys the correct meaning.

Author: It does now, thanks!

{{< /warning >}}

## Limitations

The contents of introspection relations differ depending on the selected cluster and replica.
As a consequence, you should expect the answers to the queries below to vary depending on which cluster you are working in.

> In particular, indexes and dataflows are local to a cluster, so their introspection information will vary across clusters.

Author: Most of what was previously here now lives in the `mz_internal` docs for introspection sources, as I think it should. It's probably still a good idea to point out the most important pitfalls here, as users wanting to troubleshoot something might not think to click through to the `mz_internal` docs.

Reviewer: @teskje Could you point out a "top" pitfall and we can document it here?

take the largest total amount of time.

```sql
-- Extract raw elapsed time information, by worker
SELECT mdo.id, mdo.name, mdo.worker_id, mse.elapsed_ns
-- Extract raw elapsed time information, summed across workers
SELECT mdo.id, mdo.name, mse.elapsed_ns
FROM mz_internal.mz_scheduling_elapsed AS mse,
     mz_internal.mz_dataflow_operators AS mdo
WHERE
    mse.id = mdo.id AND
    mse.worker_id = mdo.worker_id
WHERE mse.id = mdo.id
ORDER BY elapsed_ns DESC;
```

```sql
-- Extract raw elapsed time information, summed across workers
SELECT mdo.id, mdo.name, sum(mse.elapsed_ns) AS elapsed_ns
FROM mz_internal.mz_scheduling_elapsed AS mse,
-- Extract raw elapsed time information, by worker
SELECT mdo.id, mdo.name, mse.worker_id, mse.elapsed_ns
FROM mz_internal.mz_scheduling_elapsed_per_worker AS mse,
     mz_internal.mz_dataflow_operators AS mdo
WHERE
    mse.id = mdo.id AND
    mse.worker_id = mdo.worker_id
GROUP BY mdo.id, mdo.name
WHERE mse.id = mdo.id
ORDER BY elapsed_ns DESC;
```

<!-- mz_raw_compute_operator_durations is not available yet. -->

> <!-- ### Why is Materialize unresponsive for seconds at a time?

Author: I am not entirely sure why this section was commented out, but the query works (after adjusting) so it's probably fine to include it again. (cc @morsapaes)

and incriminate the subject.

```sql
-- Extract raw scheduling histogram information, summed across workers.
SELECT mdo.id, mdo.name, mcodh.duration_ns, mcodh.count
FROM mz_internal.mz_compute_operator_durations_histogram AS mcodh,
     mz_internal.mz_dataflow_operators AS mdo
WHERE mcodh.id = mdo.id
ORDER BY mcodh.duration_ns DESC;
```

```sql
-- Extract raw scheduling histogram information, by worker.
SELECT mdo.id, mdo.name, mcodh.worker_id, mcodh.duration_ns, mcodh.count
FROM mz_internal.mz_compute_operator_durations_histogram_per_worker AS mcodh,
     mz_internal.mz_dataflow_operators AS mdo
WHERE mcodh.id = mdo.id
ORDER BY mcodh.duration_ns DESC;
```

## Why is Materialize using so much memory?

> The majority of Materialize's memory use is taken up by
> [arrangements](/overview/arrangements/), which
> are differential dataflow structures that maintain indexes for data
> as it changes. These queries extract the numbers of records and

Reviewer: ```suggestion
Differential data flow structures called [arrangements](/overview/arrangements)
take up most of Materialize's memory use. Arrangements maintain indexes for data
as it changes. These queries extract the numbers of records and
```
